hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java

```diff
@@ -362,28 +361,33 @@ private HoodieMetadataColumnStats combineColumnStatsMetadatat(HoodieMetadataPayl
     return filesystemMetadata.entrySet().stream().filter(e -> e.getValue().getIsDeleted() == isDeleted);
   }
 
-  private Map<String, HoodieMetadataFileInfo> combineFilesystemMetadata(HoodieMetadataPayload previousRecord) {
+  private Map<String, HoodieMetadataFileInfo> combineFileSystemMetadata(HoodieMetadataPayload previousRecord) {
     Map<String, HoodieMetadataFileInfo> combinedFileInfo = new HashMap<>();
+
+    // First, add all files listed in the previous record
     if (previousRecord.filesystemMetadata != null) {
       combinedFileInfo.putAll(previousRecord.filesystemMetadata);
     }
 
+    // Second, merge in the files listed in the new record
     if (filesystemMetadata != null) {
-      filesystemMetadata.forEach((filename, fileInfo) -> {
-        // If the filename wasnt present then we carry it forward
-        if (!combinedFileInfo.containsKey(filename)) {
-          combinedFileInfo.put(filename, fileInfo);
-        } else {
-          if (fileInfo.getIsDeleted()) {
-            // file deletion
-            combinedFileInfo.remove(filename);
-          } else {
-            // file appends.
-            combinedFileInfo.merge(filename, fileInfo, (oldFileInfo, newFileInfo) -> {
-              return new HoodieMetadataFileInfo(oldFileInfo.getSize() + newFileInfo.getSize(), false);
-            });
-          }
-        }
+      validatePayload(type, filesystemMetadata);
+
+      filesystemMetadata.forEach((key, fileInfo) -> {
+        combinedFileInfo.merge(key, fileInfo,
+            // Combine previous record w/ the new one, new records taking precedence over
+            // the old one
+            //
+            // NOTE: That if previous listing contains the file that is being deleted by the tombstone
+            //       record (`IsDeleted` = true) in the new one, we simply delete the file from the resulting
+            //       listing as well as drop the tombstone itself.
+            //       However, if file is not present in the previous record we have to persist tombstone
+            //       record in the listing to make sure we carry forward information that this file
+            //       was deleted. This special case could occur since the merging flow is 2-stage:
+            //          - First we merge records from all of the delta log-files
+            //          - Then we merge records from base-files with the delta ones (coming as a result
+            //          of the previous step)
+            (oldFileInfo, newFileInfo) -> newFileInfo.getIsDeleted() ? null : newFileInfo);
```

Comment:
don't we need to take the max of either of sizes?

Reply:
@prashantwason : aren't there any multi-writer scenarios, where a new commit could report a file w/ lower size compared to a previous commit.

Reply:
@nsivabalan i don't think this is possible, since we never delete in-place from files.

Reply:
ok, let me take a stab to see if this is a valid scenario. this could be applicable only for hdfs.

Let's say there is a concurrent write to a log file by two diff writers. one of them is doing a rollback and another is appending a log block. Let's say writer1 (who is doing a rollback) updates the log file first and gets size as 200 may be. And later writer2 appends to same log file and gets size as 300. Even though the order in which these writers appended to file could be writer1 followed by writer2. Its not guaranteed that the same order will be maintained when they reach metadata table. So, due to various reasons writer2 could complete its write earlier and could apply its changes to metadata first before writer1. 

wouldn't we revert to a smaller size in this case when writer1 updates metadata table w/ 200 as size.

Reply:
You've brought up interesting point. First of all, MT updates and actual FS changes are actually part of the `commit` operation which is atomic (as in: it's performed under lock). 

But if extend your train of thought, and assume for a moment that it could be possible that MT updates and FS operations aren't atomic, we actually _will not be able_ to guarantee consistency b/w actual state of the FS and MT, b/c at any point MT operations (accompanying actual FS ops) could get re-ordered which practically would entail that we will potentially have incorrect MT state b/c "merge" of MT records is not a commutative operation (b/c it inherently relies on ordering for deletes). Example of such inconsistency would be "resurrecting file" where (in order)

1. File F is Appended (A_1)
2. File F is Deleted (D_1)
3. MT record (deleted=true) is added (for D_1)
4. MT record (size=..., deleted=false) is added (for A_1)

Therefore, we will have to provide for such atomicity (which we do)

Reply:
I agree. but the scenario which you quoted are unlikely to happen in general.
a. since cleaner will not touch latest file slice where regular writes go into (MOR table).
b. incase of COW rollback, there is no log file concept and hence a rollback and a commit touching the same file is very unlikely. 
So, two concurrent writes where one of them deletes while the other updates or adds a new file : I can't think of a valid scenario. may be there is, we need to think hard to see if we can come up w/ one.  

But the scenario of rollback and concurrent updates to same file slice might happen in MOR table. 

also, the fix that is being considered may not cause any regression and could possibly help thwart some of the scenarios. i.e taking the max file size rather than latest file size. Do you see any issues specifically to go w/ max of file sizes in merge function? Atleast for the rollback and regular writer in MOR use-case, we have to make the fix. I don't think we can ignore the use-case.

Reply:
@nsivabalan we're not ignoring the use-case. My point was that FS changes and MT updates **have to** be atomic. Otherwise we can't guarantee MT consistency. They're atomic currently, so the use-case you've described is impossible.

Reply:
Chatted offline w/ @nsivabalan: 

	- We can’t assume that MT update records will be ordered the same 
	way as actual FS operations (these are not atomic)
		- AI: MT record merging should be a commutative operation (not 
		assuming the records ordering)
			- This is possible for file-sizes
			- This is not possible for deletes. However we’re assuming 
			that the case of concurrent write and deletion of the same 
			file is not possible
				- This would only be possible with concurrent upset and 
				rollback operation (affecting the same log-file), which 
				is implausible, b/c either of have to be true:
					- We’re appending to failed log-file (then the other 
					writer is trying to rollback it concurrently, before 
					it’s own write)
					- Rollback (of completed instant) is running 
					concurrently with append (meaning that restore is 
					running concurrently with a write, which shouldn’t 
					occur see below)
	- Rollback should not be used to rollback completed instants (except 
	in the Restore use-case)
		- AI: We need to guard for that
	- Restore can NOT be run concurrently with any writes
		- AI: We need to guard for that (for ex, we can take a look for the 
		whole duration of the Restore)

Reply:
thanks a lot for capturing the gist. So, I assume you plan to fix the merge function to do max(old size, new size).